[RISC-V][LLD] Add Support for RISC-V TLSDESC Relocations

ilovepi opened the pull request:
This patch adds basic support to LLD for the TLSDESC relocations outlined in new https://github.com/riscv-non-isa/riscv-elf-psabi-doc/pull/373.

ilovepi committed:
[RISCV] Support RISC-V TLSDESC in LLD

This patch adds support for RISC-V TLSDESC relocations, as described in
https://github.com/riscv-non-isa/riscv-elf-psabi-doc/pull/373.

It does not attempt to handle relaxation for these cases, which will be
handled separately.

ilovepi committed:
[RISCV] Support Global Dynamic TLSDESC in the RISC-V backend

This patch adds basic TLSDESC support for the global dynamic case in the
RISC-V backend by adding new relocation types for TLSDESC, as prescribed
in https://github.com/riscv-non-isa/riscv-elf-psabi-doc/pull/373.

We also add a new pseudo instruction to simplify code generation.

Possible improvements for the local dynamic case will be addressed in separate
patches.

The current implementation is only enabled when passing the
-riscv-enable-tlsdesc flag.

rui314 commented on lld/ELF/Writer.cpp:
```diff
@@ -1897,7 +1897,8 @@ template <class ELFT> void Writer<ELFT>::finalizeSections() {
     }
 
-    if (config->emachine == EM_386 || config->emachine == EM_X86_64) {
+    if (config->emachine == EM_386 || config->emachine == EM_X86_64 ||
+        config->emachine == EM_RISCV) {
       // On targets that support TLSDESC, _TLS_MODULE_BASE_ is defined in such a
```
I believe lld supports TLSDESC for ARM64 because it's the default TLS access model on ARM64. The comment here says that `_TLS_MODULE_BASE_` is defined for targets that support TLSDESC. But the `if` condition includes only i386 and x86-64. It doesn't seem correct. Not directly related to this change, though.

MaskRay commented:
(will be out of town for ~10 days)

(I did want to implement the TLSDESC stuff myself so I can probably spend less time on reviewing: 
https://github.com/riscv-non-isa/riscv-elf-psabi-doc/issues/94#issuecomment-495966998)

ilovepi commented:
ping.

ilovepi committed:
!fixup Call and test checkPseudoTLSDESCCall

ilovepi committed:
fixup! clang-format

ishitatsuyuki commented:
FYI, one limitation of the current TLSDESC spec is that relaxation is a requirement for static binaries to work, since there will be no dynamic linker to fill in the relocs. The relaxation support can come in a separate PR, but it will need to come before enabling TLSDESC by default.

ilovepi committed:
Merge branch 'main' into gd-tlsdesc

ilovepi committed:
!fixup Address code differences after rebase

ilovepi committed:
!fixup Remove parsing for `la.tlsdesc`

Since this pseudo wasn't in the psABI, we shouldn't be emitting it
anywhere, and having the AsmParser handle it is a challenge without
introducing more operands, since it would need to know which registers
to use or have a way to scavenge registers. For now, we can leave the
pseudo-instruction as an implementation detail, with the understanding
that it shouldn't be emitted into the assembly output.

ilovepi committed:
!fixup Add TODO to remove the cl::opt option

ilovepi committed:
!fixup Rename getGeneralDynamicTLSDescAddr to getTLSDescAddr

ilovepi committed:
Merge branch 'main' into gd-tlsdesc

ilovepi committed:
Don't set isAsmParserOnly bits in PseudoLA_TLSDESC

ilovepi committed:
Add back RUN lines to test removed in old rebase

ilovepi committed:
Test tlsdesc relocations in MC layer

ilovepi committed:
Add test for invalid TLSDESC relocations

ilovepi committed:
Remove trailing whitespace

ilovepi committed:
Update error string to match checks

ilovepi committed:
Fix documentation for fixup_riscv_tlsdesc_load_lo12

ilovepi committed:
Run clang-format on patch

ilovepi committed:
Add checks both local and global accesses in test

ilovepi committed:
Merge branch 'main' into lld-tlsdesc

MaskRay commented:
The base branch can be edited if you click "Edit" near the title, which will help reveal the lld side changes...

ilovepi commented:
> The base branch can be edited if you click "Edit" near the title, which will help reveal the lld side changes...

hmm, I was hoping that would work, but the base branch is in my fork. I think to make it work correctly I'd have to be using the "user" branches in the main repository. 

Maybe its better to re-open both of them using `spr`? That still seems bad, but it will probably make everyone a bit happier. I'll look into this a bit more, because I agree, its very hard to understand what is intended to be reviewed here vs what's been included from the base patch.

ilovepi commented:
@MaskRay , do you think this is worth abandoning in favor of https://github.com/llvm/llvm-project/pull/77516? If so, I guess I'd also need to abandon https://github.com/llvm/llvm-project/pull/66915, which is unfortunate.

MaskRay commented:
> @MaskRay , do you think this is worth abandoning in favor of #77516? If so, I guess I'd also need to abandon #66915, which is unfortunate.

You may abandon this in favor of #77516. For the LLVM patch, just use #66915 as it contains a lot of discussions.
It's fine to make the lld patch not actually stacked ...

ilovepi commented:
sounds good. I'll go ahead and close this in favor of https://github.com/llvm/llvm-project/pull/77516 then, and leave the LLVM PR alone.
